README.md
# libparser
Simple libc parser to parse arguments.
Use it in your project to easy parse and use arguments.

## Platform
Linux: Build and tested
Windows: Build only
Macos: Build only

## Tests
A summary of tests is provided [here](https://github.com/osvegn/libparser/blob/main/tests/coverage_report.txt).
## How to build
### Build lib
```bash
cmake -B build
```

### Build
```bash
cmake --build build
```

### Run tests
```
cmake -B build -DTESTING=ON
cmake --build build
```

## Contribute
No specification, ask for feature by creating issue or creating pull requests with changes.

## Contact
[osvegn](thomas.prudhomme@epitech.eu)


changelog.md
# Changelog
## [V0.1.0] - 14/08/2023

First release of libparser. \
It allows to parse arguments by using option, and create an auto helper.

## Added
- Parser initialization (init_parser function)
- Parser destruction (del_parser function)
- Parse arguments logic (parse_args function)
- Add option to parser (add_option function)
- Default option '-h' or '--help' to show help.
## Changed

## Fixed


tests/test_parser.c
#include <criterion/criterion.h>
#include "parser.h"

Test(init_parser, create_empty_parser)
{
    init_parser();
    cr_assert_neq(get_parser(), 0);
    del_parser();
}

Test(init_parser, create_one_option_parser)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", "version", ""), 0);
    cr_assert_eq(get_parser()->option_count, 2);
    del_parser();
}

Test(init_parser, create_two_options_parser)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", "version", ""), 0);
    cr_assert_eq(add_option("-q", "--quit", "quit", ""), 0);
    cr_assert_eq(get_parser()->option_count, 3);
    del_parser();
}

Test(init_parser, create_parser_null_short_name)
{
    init_parser();
    cr_assert_eq(add_option(0, "--version", "version", ""), 0);
    cr_assert_eq(get_parser()->option_count, 2);
    del_parser();
}

Test(init_parser, create_parser_null_long_name)
{
    init_parser();
    cr_assert_eq(add_option("-v", 0, "version", ""), -1);
    cr_assert_eq(get_parser()->option_count, 1);
    del_parser();
}

Test(init_parser, create_parser_null_key)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", 0, ""), -1);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option)
{
    char *args[] = {"test", "--version"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_two_options)
{
    char *args[] = {"test", "--quit"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, false);
    cr_assert_eq(get_parser()->options[2].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value)
{
    char *args[] = {"test", "--version", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "vers"), 0);
    cr_assert_eq(get_parser()->options[2].found, false);
    del_parser();
}

Test(parse_args, create_parser_null_help)
{
    char *args[] = {"test", "--version", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", NULL) < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value_and_short_option)
{
    char *args[] = {"test", "-v", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "vers"), 0);
    cr_assert_eq(get_parser()->options[2].found, false);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value_using_equal)
{
    char *args[] = {"test", "--version=salut"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "salut"), 0);
    del_parser();
}

Test(parse_args, parser_multi_args_multi_options)
{
    char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-l", "--load", "load", "") < 0)
        return;
    if (add_option("-t", "--test", "test", "") < 0)
        return;
    if (add_option("-r", "--recursive", "recursive", "") < 0)
        return;
    if (add_option("-lt", "--less-than", "less-than", "") < 0)
        return;
    parse_args(6, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "\0"), 0);
    cr_assert_eq(get_parser()->options[2].found, true);
    cr_assert_eq(strcmp(get_parser()->options[2].value, ""), 0);
    cr_assert_eq(get_parser()->options[3].found, true);
    cr_assert_eq(strcmp(get_parser()->options[3].value, ""), 0);
    cr_assert_eq(get_parser()->options[4].found, false);
    cr_assert_eq(strcmp(get_parser()->options[4].value, ""), 0);
    cr_assert_eq(get_parser()->options[5].found, false);
    cr_assert_eq(strcmp(get_parser()->options[5].value, ""), 0);
    del_parser();
}

Test(get_value, test_get_value)
{
    char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_neq(get_value("version"), 0);
    del_parser();
}

Test(get_value, test_get_value_null_key)
{
    cr_assert_eq(get_value(NULL), 0);
}

Test(get_value, test_get_value_null_key_init_parser)
{
    char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value(NULL), 0);
    del_parser();
}

Test(get_value, test_get_value_empty_key_init_parser)
{
    char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value(""), 0);
    del_parser();
}

Test(get_value, test_get_value_not_found_key)
{
    char *args[] = {"test", "--load", "-r", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value("version"), 0);
    del_parser();
}

tests/unit/test_parser.c
/*
 * Filename: tests/unit/test_parser.c
 * Path: tests/unit
 * Created Date: Monday, July 17th 2023, 8:07:36 pm
 * Author: osvegn
 * 
 * Copyright (c) 2023 Your Company
 */


#include <criterion/criterion.h>
#include <criterion/redirect.h>
#include "parser.h"

void redirect_all_stdout(void)
{
    cr_redirect_stdout();
    cr_redirect_stderr();
}

Test(init_parser, create_empty_parser)
{
    init_parser();
    cr_assert_neq(get_parser(), 0);
    del_parser();
}

Test(init_parser, create_one_option_parser)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", "version", ""), 0);
    cr_assert_eq(get_parser()->option_count, 2);
    del_parser();
}

Test(init_parser, create_two_options_parser)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", "version", ""), 0);
    cr_assert_eq(add_option("-q", "--quit", "quit", ""), 0);
    cr_assert_eq(get_parser()->option_count, 3);
    del_parser();
}

Test(init_parser, create_parser_null_short_name)
{
    init_parser();
    cr_assert_eq(add_option(0, "--version", "version", ""), 0);
    cr_assert_eq(get_parser()->option_count, 2);
    del_parser();
}

Test(init_parser, create_parser_null_long_name)
{
    init_parser();
    cr_assert_eq(add_option("-v", 0, "version", ""), -1);
    cr_assert_eq(get_parser()->option_count, 1);
    del_parser();
}

Test(init_parser, create_parser_null_key)
{
    init_parser();
    cr_assert_eq(add_option("-v", "--version", 0, ""), -1);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option)
{
    const char *args[] = {"test", "--version"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_two_options)
{
    const char *args[] = {"test", "--quit"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, false);
    cr_assert_eq(get_parser()->options[2].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value)
{
    const char *args[] = {"test", "--version", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "vers"), 0);
    cr_assert_eq(get_parser()->options[2].found, false);
    del_parser();
}

Test(parse_args, create_parser_null_help)
{
    const char *args[] = {"test", "--version", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", NULL) < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value_and_short_option)
{
    const char *args[] = {"test", "-v", "vers"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-q", "--quit", "quit", "") < 0)
        return;
    parse_args(3, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "vers"), 0);
    cr_assert_eq(get_parser()->options[2].found, false);
    del_parser();
}

Test(parse_args, parser_one_arg_one_option_with_value_using_equal)
{
    const char *args[] = {"test", "--version=salut"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    parse_args(2, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "salut"), 0);
    del_parser();
}

Test(parse_args, parser_multi_args_multi_options)
{
    const char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    if (add_option("-v", "--version", "version", "") < 0)
        return;
    if (add_option("-l", "--load", "load", "") < 0)
        return;
    if (add_option("-t", "--test", "test", "") < 0)
        return;
    if (add_option("-r", "--recursive", "recursive", "") < 0)
        return;
    if (add_option("-lt", "--less-than", "less-than", "") < 0)
        return;
    parse_args(6, args);
    cr_assert_eq(get_parser()->options[1].found, true);
    cr_assert_eq(strcmp(get_parser()->options[1].value, "\0"), 0);
    cr_assert_eq(get_parser()->options[2].found, true);
    cr_assert_eq(strcmp(get_parser()->options[2].value, ""), 0);
    cr_assert_eq(get_parser()->options[3].found, true);
    cr_assert_eq(strcmp(get_parser()->options[3].value, ""), 0);
    cr_assert_eq(get_parser()->options[4].found, false);
    cr_assert_eq(strcmp(get_parser()->options[4].value, ""), 0);
    cr_assert_eq(get_parser()->options[5].found, false);
    cr_assert_eq(strcmp(get_parser()->options[5].value, ""), 0);
    del_parser();
}

Test(get_value, test_get_value)
{
    const char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_neq(get_value("version"), 0);
    del_parser();
}

Test(get_value, test_get_value_null_key)
{
    cr_assert_eq(get_value(NULL), 0);
}

Test(get_value, test_get_value_null_key_init_parser)
{
    const char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value(NULL), 0);
    del_parser();
}

Test(get_value, test_get_value_empty_key_init_parser)
{
    const char *args[] = {"test", "--load", "-v", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value(""), 0);
    del_parser();
}

Test(get_value, test_get_value_not_found_key)
{
    const char *args[] = {"test", "--load", "-r", "--help", "coucou", "--test"};

    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(6, args);
    cr_assert_eq(get_value("version"), 0);
    del_parser();
}

Test(dump, test_dump_parser, .init = redirect_all_stdout)
{
    const char *args[] = {"test", "--version"};
    char output[2048];

    memset(output, 0, 2048);
    strcpy(output, "----------\n");
    strcat(output, "option_number:\t0\n");
    strcat(output, "short_name:\t-h\n");
    strcat(output, "long_name:\t--help\n");
    strcat(output, "key:\t\thelp\n");
    strcat(output, "value:\t\t\n");
    strcat(output, "help:\t\tShow help\n");
    strcat(output, "found:\t\tFalse\n\n");
    strcat(output, "option_number:\t1\n");
    strcat(output, "short_name:\t-v\n");
    strcat(output, "long_name:\t--version\n");
    strcat(output, "key:\t\tversion\n");
    strcat(output, "value:\t\t\n");
    strcat(output, "help:\t\t\n");
    strcat(output, "found:\t\tTrue\n\n");
    strcat(output, "----------\n");
    init_parser();
    add_option("-v", "--version", "version", "");
    parse_args(2, args);
    dump();
    cr_assert_stdout_eq_str(output);
    del_parser();
}

Test(dump, test_empty_dump, .init=redirect_all_stdout)
{
    dump();
    cr_assert_stdout_eq_str("----------\n----------\n");
}

Test(is_option_found, test_is_option_found)
{
    const char *args[] = {"test", "-v"};

    init_parser();
    if (add_option("-v", "--version", "version", "version") < 0)
        cr_assert_eq(1, 0);
    if (add_option("-r", "--recursive", "recursive", "recursive") < 0)
        cr_assert_eq(1, 0);
    parse_args(2, args);
    cr_assert_eq(is_option_found("version"), true);
    cr_assert_eq(is_option_found("recursive"), false);
    del_parser();
}

Test(is_option_found, test_is_option_found_no_parser)
{
    cr_assert_eq(is_option_found("version"), false);
}

Test(is_option_found, test_is_option_found_invalid_option)
{
    const char *args[] = {"test", "-r"};

    init_parser();
    if (add_option("-r", "--recursive", "recursive", "recursive") < 0)
        cr_assert_eq(1, 0);
    parse_args(2, args);
    cr_assert_eq(is_option_found("version"), false);
}


includes/parser.h
/*
 * Filename: includes/parser.h
 * Path: includes
 * Created Date: Friday, August 4th 2023, 1:28:12 pm
 * Author: osvegn
 * 
 * Copyright (c) 2023 Your Company
 */


#include <stdbool.h>

typedef struct option_s {
    char short_name[256];
    char long_name[512];
    char key[256];
    char value[512];
    char help[1024];
    bool found;
} option_t;

typedef struct parser_s {
    option_t *options;
    int option_count;
} parser_t;

/**
 * @author osvegn
 * @name init_parser
 * @brief Parser initializer. Must be used at the beginning to
 * setup the parser.
*/
void init_parser(void);

/**
 * @author osvegn
 * @name get_parser
 * @brief Getter which returns the instance of the parser.
 * @return const parser_t pointer.
*/
parser_t *get_parser(void);

/**
 * @author osvegn
 * @name del_parser
 * @brief Parser destructor. Must be used at the end to delete
 * the parser.
*/
void del_parser(void);

/**
 * @author osvegn
 * @name parse_args
 * @brief Argument parser. It fills defined options with
 * corresponding values.
 * @param ac int | Number of arguments.
 * @param av array of string | Arguments.
 * @return int | 0
*/
int parse_args(int ac, const char **av);

/**
 * @author osvegn
 * @name add_option
 * @brief Configure new option. It defines new available option to parser.
 * @param short_name str | Short name of the option (256).
 * (example: "-v")
 * @param long_name str | Long name of the option (512).
 * (example: "--version")
 * @param key str | Key of the option (256). (example: "version")
 * @param help str | Help message to print on help (1024).
 * (example: "Print version")
 * @return int | -1 if long_name or key is null, or if realloc failed.
 * 0 on success.
*/
int add_option(char *short_name, char *long_name, char *key, char *help);

/**
 * @author osvegn
 * @name get_value
 * @brief Get value from key.
 * @param key str | Key to get value from.
 * @return str | Value of the key.
 * 0 if key not found or if option isn't filled.
*/
char *get_value(char *key);

/**
 * @author osvegn
 * @name dump
 * @brief Dump parser. It display in a human readable way the stat of
 * the parser.
*/
void dump(void);

/**
 * @author osvegn
 * @name is_option_found
 * @brief It checks if key was found or not.
 * @param str | The option's key.
 * @param bool | true if found, false otherwise.
*/
bool is_option_found(char *key);


sources/parser.c
/*
 * Filename: sources/parser.c
 * Path: sources
 * Created Date: Friday, August 4th 2023, 1:28:12 pm
 * Author: osvegn
 * 
 * Copyright (c) 2023 Your Company
 */


#include "parser.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

static parser_t parser = {0};

void init_parser(void)
{
    parser.options = 0;
    parser.option_count = 0;
    add_option("-h", "--help", "help", "Show help");
}

parser_t *get_parser(void)
{
    return &parser;
}

void del_parser(void)
{
    if (parser.options)
        free(parser.options);
}


int add_option(char *short_name, char *long_name, char *key, char *help)
{
    if (!long_name || !key)
        return -1;
    parser.options = realloc(parser.options, sizeof(option_t) * (parser.option_count + 1));
    if (!parser.options)
        return -1;
    strncpy(parser.options[parser.option_count].long_name, long_name, 512);
    if (short_name)
        strncpy(parser.options[parser.option_count].short_name, short_name, 256);
    else
        memset(parser.options[parser.option_count].short_name, 0, 256);
    if (help)
        strncpy(parser.options[parser.option_count].help, help, 1024);
    else
        memset(parser.options[parser.option_count].help, 0, 1024);
    strncpy(parser.options[parser.option_count].key, key, 256);
    parser.options[parser.option_count].found = false;
    memset(parser.options[parser.option_count].value, 0, 512);
    parser.option_count++;
    return 0;
}

static bool is_in_parser(char *str, int *index)
{
    int len = 0;

    for ((*index) = 0; (*index) < parser.option_count; (*index)++) {
        len = strlen(parser.options[(*index)].short_name);
        if (strncmp(str, parser.options[(*index)].short_name, len) == 0)
            return true;
        len = strlen(parser.options[(*index)].long_name);
        if (strncmp(str, parser.options[(*index)].long_name, len) == 0)
            return true;
    }
    (*index) = 0;
    return false;
}

int parse_args(int ac, const char **av)
{
    int option_index = 0;

    for (int i = 1; i < ac; i++) {
        if (is_in_parser(av[i], &option_index)) {
            parser.options[option_index].found = true;
            int tmp = option_index;
            if (i + 1 < ac && !is_in_parser(av[i + 1], &tmp))
                strncpy(parser.options[option_index].value, av[i + 1], 512);
            if (av[i][strlen(parser.options[option_index].long_name)] == '=')
                strncpy(parser.options[option_index].value, av[i] + strlen(parser.options[option_index].long_name) + 1, 512);
        }
    }
    return 0;
}

static int index_from_key(char *key)
{
    int i = 0;

    if (!key)
        return -1;
    for (; i < parser.option_count; i++) {
        if (strcmp(parser.options[i].key, key) == 0)
            return i;
    }
    return -1;
}

char *get_value(char *key)
{
    int index = index_from_key(key);

    if (index < 0)
        return 0;
    if (!parser.options[index].found)
        return 0;
    return parser.options[index].value;
}

void dump(void)
{
    int i = 0;
    option_t *option = 0;

    printf("----------\n");
    while (parser.option_count > i) {
        option = &parser.options[i];
        printf("option_number:\t%d\nshort_name:\t%s\nlong_name:\t%s\nkey:\t\t%s\nvalue:\t\t%s\nhelp:\t\t%s\nfound:\t\t%s\n\n", i, option->short_name, option->long_name, option->key, option->value, option->help, (option->found) ? "True" : "False");
        i++;
    }
    printf("----------\n");
    fflush(stdout);
}

bool is_option_found(char *key)
{
    int index = index_from_key(key);

    if (index < 0)
        return false;
    return parser.options[index].found;
}
